Subject: Fix typo in libaudit requirement
Origin: v8-1-g69524aa
Upstream-Author: Nathaniel McCallum <npmccallum@redhat.com>
Date: Tue Nov 14 10:24:35 2017 -0500

--- a/configure.ac
+++ b/configure.ac
@@ -18,7 +18,7 @@
 PKG_CHECK_MODULES([jose], [jose >= 8])
 PKG_CHECK_MODULES([systemd], [systemd])
 PKG_CHECK_MODULES([dracut], [dracut])
-PKG_CHECK_MODULES([audit], [audit >> 2.7.8])
+PKG_CHECK_MODULES([audit], [audit >= 2.7.8])
 
 AC_CHECK_PROG([PWMAKE], [pwmake], [yes])
 test -n "$PWMAKE" || AC_MSG_ERROR([pwmake required!])
